src/ucp/core/ucp_device.c

```diff
@@ -121,38 +183,61 @@ ucp_device_mem_list_params_check(const ucp_device_mem_list_params_t *params,
                                RKEY, NULL);
 
         /* TODO: Delegate most of checks below to proto selection */
-        if ((rkey == NULL) || (memh == NULL)) {
-            ucs_error("element[%lu] rkey=%p, memh=%p", i, rkey, memh);
+        if (rkey == NULL) {
+            ucs_error("element[%lu] rkey is NULL", i);
             return UCS_ERR_INVALID_PARAM;
         }
 
         if (i == 0) {
-            *local_sys_dev  = memh->sys_dev;
-            *local_md_map   = memh->md_map;
-            *mem_type       = memh->mem_type;
+            if (memh != NULL) {
+                *local_sys_dev = memh->sys_dev;
+                *local_md_map  = memh->md_map;
+                *mem_type      = memh->mem_type;
+            }
+
             *rkey_cfg_index = rkey->cfg_index;
             if (*rkey_cfg_index == UCP_WORKER_CFG_INDEX_NULL) {
                 ucs_debug("invalid first rkey: cfg_index=%d", *rkey_cfg_index);
                 return UCS_ERR_INVALID_PARAM;
             }
         } else {
-            *local_md_map &= memh->md_map;
             if (rkey->cfg_index != *rkey_cfg_index) {
                 ucs_debug("mismatched rkey config index: "
                           "ucp_rkey[%lu]->cfg_index=%u cfg_index=%u",
                           i, rkey->cfg_index, *rkey_cfg_index);
                 return UCS_ERR_UNSUPPORTED;
             }
 
-            if (memh->sys_dev != *local_sys_dev) {
-                ucs_debug("mismatched local sys_dev: ucp_memh[%zu].sys_dev=%u "
-                          "first_sys_dev=%u",
-                          i, memh->sys_dev, *local_sys_dev);
-                return UCS_ERR_UNSUPPORTED;
+            if (memh != NULL) {
+                if (*local_sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) {
+                    *local_sys_dev = memh->sys_dev;
+                    *local_md_map  = memh->md_map;
+                    *mem_type      = memh->mem_type;
+                } else {
+                    *local_md_map &= memh->md_map;
+                    if (memh->sys_dev != *local_sys_dev) {
```

Comment:
The loop contains too many if/else blocks.
I suggest to split the loop: one checks `memh`, another checks `rkey`; since the checks are independent. And put them into separate functions.